torch/_dynamo/variables/builtin.py

```diff
@@ -559,6 +559,35 @@ def call_enumerate(self, tx, *args):
             ]
             return variables.TupleVariable(items, **options)
 
+    def _call_operator_builtin(self, tx, op_name: str, *args, **kwargs):
+        options = VariableTracker.propagate(self, args)
+        fn_ = getattr(operator, op_name)
+
+        # This should really be refactored.
+        # builtins like + and - can return a ton of different types,
+        # depending on the types of the inputs.
+        # Right now, TensorVariable.create() handles a subset of these cases,
+        # but not all of them.
+        # For example, different cases that we care about include
+        # when the output type is a:
+        # - ConstantVariable (handled by ConstantVariable.call_method)
+        # - DynamicShapesVariable (handled by TensorVariable.create)
+        # - TensorVariable (handled by TensorVariable.create)
+        if all(isinstance(x, variables.ConstantVariable) for x in args):
+            return args[0].call_method(tx, f'__{op_name}__', args[1:], kwargs)
```

Comment:
This got more models to pass, but feels in need of a refactor.

**(1)** Functions like `operator.add` can work for a large variety of types (ints, SymInts, Tensors, etc), and we have a closed set of `VariableTracker` subclasses that this function can return in dynamo (`ConstantVariable`, `DynamicShapeVariable`, `TensorVariable`, not sure if there are others).

One general dynamo question that I have is that traced functions that can operate on many types, what do we want the logic to look like? In principle, what it seems like we want is:

- trace the underlying function with proxies to get a proxy output
- inspect the type of the proxy output (int, SymInt, Tensor, etc)
- figure out what VariableTracker sublass to return based on the output type

And I think... `TensorVariable.create()` sort of does this, but is incomplete? e.g. it doesn't handle int/float outputs (I had to special case for ConstantVariables here).

@voznesenskym definitely interested in your opinion (or if I'm holding this wrong and/or this is already being done elsewhere).



**(2)** A more principled way to map all builtin ops `+` `-` `/` `*` into their `operator.foo` equivalents in dynamo. Maybe we should just keep around a list of binary-op-builtins that need this behavior.

Reply:
Separately - we have a one-to-many relationship betweena `VariableTracker` subclass and the type of the object(s) that it holds (inside of its `proxy.node.meta['example_value']`):

`DynamicShapeVariable` -> `torch.Symint`, `torch.SymFloat`
`ConstantVariable` -> `int`, `float`, idk what else
`TensorVariable` -> torch.Tensor

It would be nice enforce that invariant a bit more strictly.

Reply:
jansel: push this method into call_method (TODO: explain more). Can also try call method, and if it is not implemented, do something else.

Reply:
or just implement call method on symint so it knows to do the swapping

Reply:
but if it affects a lot of methods, if it already raises, use exceptions, catch it and handle it here

Reply:
> or just implement call method on symint so it knows to do the swapping

To be clear, this suggestion is to continue tracing `__add__` and friends instead of `operator.add`, but manually implement the "check for `NotImplemented` and swizzle arguments" logic in dynamo?

And the main benefit of that I as I see it is we can continue just use `arg.call_method()` in all cases and it'll do the correct thing.

I guess we're relying on the invariant that `typeA.__foo__(typeB)` will either always return something of `typeA`, or return `NotImplemented`. Which, sounds fair to assume? At least for the types that dynamo cares about.